### DISCUSSION TITLE
kqcircuits.util.macro_prepare package missing

### DISCUSSION THREAD

**Discussion by Qvadi** [iqm-finland/KQCircuits#60] [Q&A] [answered]
Hi, 
When running KLayout_macro, I get this error saying kqcircuits.util.macro_prepare package missing. This doesnt happen if I just simply download the KQCircuits package from the package manager. But if I git clone KQCircuits, then I get this error. 
I dont see this util macro on the API documentation page as well. Any idea how I can fix this?

**Comment by qpavsmi** [comment]
Hi,
Try running if you haven't `python setup_within_klayout.py` in the folder where you cloned KQCircuits, this should link all of the required kqcircuits directories to KLayout. Relevant chapter from docs here: https://iqm-finland.github.io/KQCircuits/developer/setup.html

Have you removed KQCircuits from KLayout's package manager before cloning from GitHub? It's possible that two parallel KQC installations interfere.

If these don't work, when you reply, could you also tell us what is the os in your system and the version of the KLayout? Also if you could specify where have you located `KLayout_macro`, this doesn't seem to be a macro that we provide with KQCircuits but I could be mistaken

**Comment by qpavsmi** [comment]
After investigating further this does seem to be a python file that we have removed at some point from the repository, but perhaps the SALT package still has it since we haven't updated the SALT package in a while. If you could send the contents of the `KLayout_macro` and we could see what we can do to make it work with current version of KQCircuits

**Comment by Qvadi** [comment]
Hi, 
thanks for the reply. I did both:
1. Installed the package uisng salt only. I did not get this error.
2. Uninstalled the salt package, and git cloned the package. Then I see this error. The contents of the Klayout_macro as as follows:
3. I did not get any errors using package installation using both methods.

# This file is to be run in KLayout Macro Development

import pya
import kqcircuits.util.macro_prepare as macroprep
from kqcircuits.chips.quality_factor import QualityFactor

(layout, layout_view, cell_view) = macroprep.prep_empty_layout()
top_cell = layout.create_cell("Top Cell")
cell_view.cell_name=top_cell.name     # Shows the new cell

# copy your resonator parameters here
res_lengths = [4511.28696175443, 4713.33669564261, 4932.686872240668, 5174.906049896605, 5439.156417772476, 5734.9182128497205]
n_fingers = [4, 4, 4, 4, 4, 4]
type_coupler = ['gap', 'gap', 'gap', 'gap', 'gap', 'gap']
l_fingers = [30.44618557383356, 51.255768510788975, 24.19872442723669, 42.262019745974065, 19.68406910466407, 36.38059885651449]
res_a = [20.0, 20.0, 10.0, 10.0, 5.0, 5.0]
res_b = [12.0, 12.0, 6.0, 6.0, 3.0, 3.0]
a = 10.0
b = 6.0

# create chip and insert it to the top_cell
chip_size = 10000
chip = QualityFactor.create(layout, box=pya.DBox(pya.DPoint(0,0),pya.DPoint(chip_size,chip_size)), 
    res_lengths=res_lengths,
    n_fingers=n_fingers, type_coupler=type_coupler, l_fingers=l_fingers,
    res_a=res_a, res_b=res_b, a=a, b=b,
    n_ab=[0]*len(res_lengths), res_term=["galvanic"]*len(res_lengths),
    res_beg=["galvanic"]*len(res_lengths), tl_airbridges=False
)
top_cell.insert(pya.DCellInstArray(chip.cell_index(), pya.DTrans(pya.DPoint(0, 0))))

I can see most of the libraries (more from git clone than salt, some are missing in salt). I can place elements in klayout using both methods, but the macros dont run because of the missing package in the second method. The first method also has some errors but maybe we can take it up on another thread.

**Reply by qpavsmi** [reply, marked as answer]
Hi, thanks for the details!

I think the following code accomplishes the same thing you intend to do:

```
import pya
from kqcircuits.klayout_view import KLayoutView
from kqcircuits.chips.quality_factor import QualityFactor

view = KLayoutView()
```
Then you can access `layout_view`, `cell_view`, `layout`, `top_cell` and `active_cell` using `view.layout_view`, `view.cell_view` etc. You can read more on [KLayoutView API](https://iqm-finland.github.io/KQCircuits/api/kqcircuits.klayout_view.html) and a tutorial on writing [macros](https://iqm-finland.github.io/KQCircuits/user_guide/macro_workflow.html)

For most purposes I would recommend the second installation configuration: No SALT package, git clone KQCircuits and run `setup_within_klayout.py`. SALT package is offered as a way to quickly try out KQCircuits or to demo it in-location.

**Reply by Qvadi** [reply]
When I add elements into the layout using the GUI, is it possible to record this as a macro while performing these tasks? Then it becomes much easier to see how things are being built inside the code and build on top of that.

**Reply by qpavsmi** [reply]
`scripts/macros/export/export_cells_as_code.lym` might be the macro you are looking for. It will look at the layout and generate python code that if run as macro, should return the same geometry that you have placed by hand. Have a try and let me know if there are issues with it.

But when it comes to having macros work with git cloned installation of KQCircuits, can we consider this discussion thread to be closed? You are always welcome to open a new discussion thread as the topic changes

**Reply by Qvadi** [reply]
Yes! thank you.
